File: .claude/skills/wiki/scripts/mermaid_lint.py
#!/usr/bin/env python3
"""Mermaid diagram syntax validation for wiki pages.

Extracts fenced ``mermaid`` code blocks, validates structural syntax
(diagram type, balanced brackets), and reports errors as lint warnings.

Usage as a library:
    from mermaid_lint import extract_mermaid_blocks, validate_block, lint_page
    blocks = extract_mermaid_blocks(content)
    errors = validate_block(block_content)
    issues = lint_page("wiki/auth/session.md")

Usage as a script:
    python mermaid_lint.py --page wiki/auth/session.md
    python mermaid_lint.py --wiki-root /path/to/wiki
"""
from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

# ── Constants ───────────────────────────────────────────────────────

VALID_DIAGRAM_TYPES = {
    "erDiagram", "sequenceDiagram", "graph", "flowchart",
    "classDiagram", "stateDiagram", "gantt", "pie", "gitgraph",
}

_MERMAID_BLOCK_RE = re.compile(r"```mermaid\s*\n(.*?)```", re.DOTALL)

_BRACKET_PAIRS = {"[": "]", "{": "}", "(": ")"}

# Diagram types where { } are used in relationship syntax, not as block delimiters
_RELATIONSHIP_BRACKET_TYPES = {"erDiagram"}

# ── Library API ─────────────────────────────────────────────────────


def extract_mermaid_blocks(content: str) -> list[dict]:
    """Extract all fenced mermaid code blocks from markdown content.

    Returns a list of dicts, each with:
    - ``line``: 1-indexed line number of the opening fence
    - ``content``: the block content (between fences)
    - ``diagram_type``: first keyword (e.g. ``erDiagram``, ``graph``)
    """
    blocks = []
    for match in _MERMAID_BLOCK_RE.finditer(content):
        line_num = content[:match.start()].count("\n") + 1
        block_content = match.group(1).strip()
        # Diagram type is the first non-empty token
        first_line = block_content.split("\n")[0].strip() if block_content else ""
        diagram_type = first_line.split()[0] if first_line else ""
        blocks.append({
            "line": line_num,
            "content": block_content,
            "diagram_type": diagram_type,
        })
    return blocks


def validate_block(block: str) -> list[str]:
    """Validate a mermaid block's structural syntax.

    Returns a list of error messages.  Empty list means valid.
    """
    block = block.strip()
    errors = []

    if not block:
        errors.append("Empty mermaid block")
        return errors

    # Check diagram type
    first_line = block.split("\n")[0].strip()
    diagram_type = first_line.split()[0] if first_line else ""

    if diagram_type not in VALID_DIAGRAM_TYPES:
        errors.append(
            f"Unrecognized diagram type '{diagram_type}'. "
            f"Expected one of: {', '.join(sorted(VALID_DIAGRAM_TYPES))}"
        )

    # Check balanced brackets (skip for diagram types that use { } in syntax)
    if diagram_type not in _RELATIONSHIP_BRACKET_TYPES:
        stack = []
        for char in block:
            if char in _BRACKET_PAIRS:
                stack.append(char)
            elif char in _BRACKET_PAIRS.values():
                if not stack:
                    errors.append(f"Unmatched closing bracket '{char}'")
                    break
                opener = stack.pop()
                if _BRACKET_PAIRS[opener] != char:
                    errors.append(
                        f"Mismatched brackets: '{opener}' opened but '{char}' found"
                    )
                    break
        if stack:
            unclosed = "".join(stack)
            errors.append(f"Unclosed brackets: {unclosed}")

    return errors


def lint_page(page_path: str) -> list[dict]:
    """Lint all mermaid blocks in a single wiki page.

    Returns a list of issue dicts for blocks with errors:
    ``{"page": str, "line": int, "errors": [str]}``
    """
    content = Path(page_path).read_text()
    blocks = extract_mermaid_blocks(content)
    issues = []
    for block in blocks:
        errors = validate_block(block["content"])
        if errors:
            issues.append({
                "page": page_path,
                "line": block["line"],
                "errors": errors,
            })
    return issues


def lint_wiki_mermaid(wiki_root: str) -> list[dict]:
    """Lint all mermaid blocks in all wiki pages.

    Globs ``wiki/**/*.md`` under *wiki_root*.
    """
    wiki_dir = Path(wiki_root) / "wiki"
    if not wiki_dir.exists():
        return []
    issues = []
    for page in sorted(wiki_dir.rglob("*.md")):
        issues.extend(lint_page(str(page)))
    return issues


# ── CLI ─────────────────────────────────────────────────────────────


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Validate Mermaid diagram syntax in wiki pages."
    )
    parser.add_argument("--page", help="Lint a single page")
    parser.add_argument("--wiki-root", help="Lint all pages in wiki/")
    return parser


def main(argv: list[str] = None) -> None:
    parser = _build_parser()
    args = parser.parse_args(argv)

    if args.wiki_root:
        issues = lint_wiki_mermaid(args.wiki_root)
    elif args.page:
        issues = lint_page(args.page)
    else:
        parser.print_help()
        return

    print(json.dumps(issues, indent=2))


if __name__ == "__main__":
    main()

File: .claude/skills/wiki/scripts/mermaid_lint.js
#!/usr/bin/env node
/**
 * Mermaid diagram syntax validation for wiki pages.
 *
 * Extracts fenced ``mermaid`` code blocks, validates structural syntax
 * (diagram type, balanced brackets), and reports errors as lint warnings.
 *
 * Usage as a library:
 *     import { extractMermaidBlocks, validateBlock, lintPage } from "./mermaid_lint.js";
 *     const blocks = extractMermaidBlocks(content);
 *     const errors = validateBlock(blockContent);
 *     const issues = lintPage("wiki/auth/session.md");
 *
 * Usage as a script:
 *     node mermaid_lint.js --page wiki/auth/session.md
 *     node mermaid_lint.js --wiki-root /path/to/wiki
 *
 * This is a TypeScript port of mermaid_lint.py; behaviour and CLI output
 * match the Python reference byte-for-byte for the same inputs.
 */
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
// ── Constants ───────────────────────────────────────────────────────
export const VALID_DIAGRAM_TYPES = new Set([
    "erDiagram",
    "sequenceDiagram",
    "graph",
    "flowchart",
    "classDiagram",
    "stateDiagram",
    "gantt",
    "pie",
    "gitgraph",
]);
/**
 * Matches Python's `re.compile(r"```mermaid\s*\n(.*?)```", re.DOTALL)`.
 * Constructed dynamically to avoid closing the ``` inside a JS template.
 * `s` flag = DOTALL; `g` flag = iterate all matches.
 */
const _MERMAID_BLOCK_RE = new RegExp("```mermaid\\s*\\n([\\s\\S]*?)```", "g");
const _BRACKET_PAIRS = {
    "[": "]",
    "{": "}",
    "(": ")",
};
const _BRACKET_OPENERS = new Set(["[", "{", "("]);
const _BRACKET_CLOSERS = new Set(["]", "}", ")"]);
/** Diagram types where `{ }` are used in relationship syntax, not as block
 *  delimiters — bracket balancing is skipped for these. */
const _RELATIONSHIP_BRACKET_TYPES = new Set(["erDiagram"]);
/**
 * Extract all fenced mermaid code blocks from markdown content.
 *
 * Returns a list of dicts, each with:
 * - `line`: 1-indexed line number of the opening fence
 * - `content`: the block content (between fences)
 * - `diagram_type`: first keyword (e.g. `erDiagram`, `graph`)
 */
export function extractMermaidBlocks(content) {
    const blocks = [];
    // Reset the regex's internal state before scanning — `g` flag is stateful.
    _MERMAID_BLOCK_RE.lastIndex = 0;
    let match;
    while ((match = _MERMAID_BLOCK_RE.exec(content)) !== null) {
        // Python: content[:match.start()].count("\n") + 1
        const startIdx = match.index;
        let newlines = 0;
        for (let i = 0; i < startIdx; i++) {
            if (content[i] === "\n")
                newlines++;
        }
        const lineNum = newlines + 1;
        const captured = match[1] ?? "";
        const blockContent = captured.trim();
        // Diagram type is the first non-empty token
        const firstLine = blockContent
            ? (blockContent.split("\n")[0] ?? "").trim()
            : "";
        const diagramType = firstLine
            ? (firstLine.split(/\s+/).filter((s) => s.length > 0)[0] ?? "")
            : "";
        blocks.push({
            line: lineNum,
            content: blockContent,
            diagram_type: diagramType,
        });
    }
    return blocks;
}
/**
 * Validate a mermaid block's structural syntax.
 *
 * Returns a list of error messages. Empty list means valid.
 */
export function validateBlock(block) {
    const trimmed = block.trim();
    const errors = [];
    if (!trimmed) {
        errors.push("Empty mermaid block");
        return errors;
    }
    // Check diagram type
    const firstLine = (trimmed.split("\n")[0] ?? "").trim();
    const diagramType = firstLine
        ? (firstLine.split(/\s+/).filter((s) => s.length > 0)[0] ?? "")
        : "";
    if (!VALID_DIAGRAM_TYPES.has(diagramType)) {
        const sorted = [...VALID_DIAGRAM_TYPES].sort();
        errors.push(`Unrecognized diagram type '${diagramType}'. ` +
            `Expected one of: ${sorted.join(", ")}`);
    }
    // Check balanced brackets (skip for diagram types that use { } in syntax)
    if (!_RELATIONSHIP_BRACKET_TYPES.has(diagramType)) {
        const stack = [];
        let stopped = false;
        // TODO: balance check ignores quote context; `A["{test}"]` would miscount.
        // Preserved parity with mermaid_lint.py — both have this pre-existing limitation.
        for (const ch of trimmed) {
            if (_BRACKET_OPENERS.has(ch)) {
                stack.push(ch);
            }
            else if (_BRACKET_CLOSERS.has(ch)) {
                if (stack.length === 0) {
                    errors.push(`Unmatched closing bracket '${ch}'`);
                    stopped = true;
                    break;
                }
                const opener = stack.pop();
                if (_BRACKET_PAIRS[opener] !== ch) {
                    errors.push(`Mismatched brackets: '${opener}' opened but '${ch}' found`);
                    stopped = true;
                    break;
                }
            }
        }
        if (!stopped && stack.length > 0) {
            const unclosed = stack.join("");
            errors.push(`Unclosed brackets: ${unclosed}`);
        }
    }
    return errors;
}
/**
 * Lint all mermaid blocks in a single wiki page.
 *
 * Returns a list of issue dicts for blocks with errors:
 * `{page, line, errors}`
 */
export function lintPage(pagePath) {
    const content = fs.readFileSync(pagePath, { encoding: "utf-8" });
    const blocks = extractMermaidBlocks(content);
    const issues = [];
    for (const block of blocks) {
        const errors = validateBlock(block.content);
        if (errors.length > 0) {
            issues.push({
                page: pagePath,
                line: block.line,
                errors,
            });
        }
    }
    return issues;
}
/** Recursively enumerate all `.md` files beneath `root`, sorted lexicographically
 *  to match Python's `sorted(Path(wiki_dir).rglob("*.md"))`. */
function rglobMarkdown(root) {
    const out = [];
    const stack = [root];
    while (stack.length > 0) {
        const dir = stack.pop();
        if (dir === undefined)
            continue;
        let entries;
        try {
            entries = fs.readdirSync(dir, { withFileTypes: true });
        }
        catch {
            continue;
        }
        for (const entry of entries) {
            const full = path.join(dir, entry.name);
            if (entry.isDirectory()) {
                stack.push(full);
            }
            else if (entry.isFile() && full.endsWith(".md")) {
                out.push(full);
            }
        }
    }
    out.sort();
    return out;
}
/**
 * Lint all mermaid blocks in all wiki pages.
 *
 * Globs `wiki/**\/*.md` under `wikiRoot`.
 */
export function lintWikiMermaid(wikiRoot) {
    const wikiDir = path.join(wikiRoot, "wiki");
    if (!fs.existsSync(wikiDir)) {
        return [];
    }
    const issues = [];
    for (const page of rglobMarkdown(wikiDir)) {
        issues.push(...lintPage(page));
    }
    return issues;
}
/**
 * Hand-rolled argparse-equivalent. Accepts `--flag value` or `--flag=value`
 * and the `-h`/`--help` flag. Unknown flags throw.
 */
function parseArgs(argv) {
    const out = {};
    let i = 0;
    while (i < argv.length) {
        const a = argv[i];
        if (a === undefined) {
            i++;
            continue;
        }
        if (a === "-h" || a === "--help") {
            out.help = true;
            i++;
            continue;
        }
        let name;
        let value;
        if (a.startsWith("--")) {
            const eq = a.indexOf("=");
            if (eq >= 0) {
                name = a.slice(2, eq);
                value = a.slice(eq + 1);
                i++;
            }
            else {
                name = a.slice(2);
                value = argv[i + 1];
                i += 2;
            }
        }
        else {
            throw new Error(`unrecognized argument: ${a}`);
        }
        switch (name) {
            case "page":
                out.page = value ?? "";
                break;
            case "wiki-root":
                out.wikiRoot = value ?? "";
                break;
            default:
                throw new Error(`unrecognized argument: --${name}`);
        }
    }
    return out;
}
const HELP_TEXT = `usage: mermaid_lint.js [-h] [--page PAGE] [--wiki-root WIKI_ROOT]

Validate Mermaid diagram syntax in wiki pages.

options:
  -h, --help            show this help message and exit
  --page PAGE           Lint a single page
  --wiki-root WIKI_ROOT
                        Lint all pages in wiki/
`;
export function main(argv = process.argv.slice(2)) {
    let args;
    try {
        args = parseArgs(argv);
    }
    catch (e) {
        process.stderr.write(`${e.message}\n`);
        return 2;
    }
    if (args.help) {
        process.stdout.write(HELP_TEXT);
        return 0;
    }
    let issues;
    if (args.wikiRoot) {
        issues = lintWikiMermaid(args.wikiRoot);
    }
    else if (args.page) {
        issues = lintPage(args.page);
    }
    else {
        process.stdout.write(HELP_TEXT);
        return 0;
    }
    process.stdout.write(JSON.stringify(issues, null, 2) + "\n");
    return 0;
}
// CLI entry point: run main() when this file is executed directly.
const thisFile = fileURLToPath(import.meta.url);
if (process.argv[1] && path.resolve(process.argv[1]) === thisFile) {
    process.exit(main());
}
